testsuite/forge/src/backend/k8s/swarm.rs

```diff
@@ -203,8 +159,12 @@ impl Swarm for K8sSwarm {
             .map(|v| v as &mut dyn Validator)
     }
 
-    fn upgrade_validator(&mut self, _id: PeerId, _version: &Version) -> Result<()> {
-        todo!()
+    fn upgrade_validator(&mut self, id: PeerId, version: &Version) -> Result<()> {
+        let validator = self
+            .validators
+            .get_mut(&id)
+            .ok_or_else(|| anyhow!("Invalid id: {}", id))?;
+        set_validator_image_tag(&validator.name(), &format!("{}", version), &self.helm_repo)
```

Comment:
You should probably follow the pattern set in the Local backend to have a mapping from `Version -> K8sVersion` instead of calling `.to_string` on the Version provided. The Display impl on Version was only intended to be used to allow tests to "print" out the version information, not as a mechanism for the backends to expect to get the correct image tag back out. Specifically because a test can trivially create a Version with a bogus display string causing the backend to do something unexpect (load an image we don't actually want to test for example)

Reply:
Oh it looks like you do have a mapping, then what you should be doing here is looking up the mapping and using that version instead. `self.version.get(version)`

Reply:
u are fully right, I did forget to clean these part since I did not do version at the first version. yea, it is a hashmap we can easily get the value